File: src/main/java/com/converter/api/config/AuthenticationTokenFilter.java
package com.converter.api.config;

import com.converter.api.exception.TokenException;
import com.converter.api.model.User;
import com.converter.api.repository.UserRepository;
import com.converter.api.service.TokenService;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.filter.OncePerRequestFilter;

import javax.servlet.FilterChain;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.util.NoSuchElementException;

public class AuthenticationTokenFilter extends OncePerRequestFilter {

    private final TokenService tokenService;
    private final UserRepository userRepository;

    public AuthenticationTokenFilter(TokenService tokenService, UserRepository userRepository) {
        this.tokenService = tokenService;
        this.userRepository = userRepository;
    }

    @Override
    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain)
            throws ServletException, IOException {

        String token = recoverToken(request);
        boolean isValid = tokenService.isValidToken(token);
        if (isValid) {
            autenticateUser(token);
        }

        filterChain.doFilter(request, response);
    }

    private void autenticateUser(String token) {
        Long userId = tokenService.getUserId(token);
        try {
            User user = userRepository.findById(userId).get();
            UsernamePasswordAuthenticationToken authentication = new UsernamePasswordAuthenticationToken(user, null, user.getAuthorities());
            SecurityContextHolder.getContext().setAuthentication(authentication);
        } catch (NoSuchElementException e){
            throw new TokenException();
        }
    }

    private String recoverToken(HttpServletRequest request) {
        String token = request.getHeader("Authorization");
        if (token == null || token.isEmpty() || !token.startsWith("Bearer ")) {
            return null;
        }

        return token.substring(7, token.length());
    }
}

File: src/main/kotlin/com/converter/api/config/AuthenticationTokenFilter.kt
package com.converter.api.config

import com.converter.api.repository.UserRepository
import org.springframework.web.filter.OncePerRequestFilter
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse
import javax.servlet.FilterChain
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.context.SecurityContextHolder
import java.util.NoSuchElementException
import com.converter.api.exception.TokenException
import com.converter.api.service.TokenService

class AuthenticationTokenFilter(
    private val tokenService: TokenService,
    private val userRepository: UserRepository
) : OncePerRequestFilter() {

    override fun doFilterInternal(
        request: HttpServletRequest,
        response: HttpServletResponse,
        filterChain: FilterChain
    ) {

        recoverToken(request)?.also {
            if (tokenService.isValidToken(it)) {
                autenticateUser(it)
            }
            filterChain.doFilter(request, response)
        }
    }

    private fun autenticateUser(token: String) =
        try {
            tokenService.getUserId(token).also {
                val user = userRepository.findById(it).get()
                val authentication = UsernamePasswordAuthenticationToken(user, null, user.authorities)
                SecurityContextHolder.getContext().authentication = authentication
            }
        } catch (e: NoSuchElementException) {
            throw TokenException()
        }


    private fun recoverToken(request: HttpServletRequest): String? {

        val token = request.getHeader("Authorization")

        return if (token == null || token.isEmpty() || !token.startsWith("Bearer ")) {
            null
        } else
            token.substring(7, token.length)
    }
}